src/Parallel/NodeLock.hpp

```diff
@@ -8,66 +8,93 @@
 #include "Utilities/Gsl.hpp"
 #include "Utilities/NoSuchType.hpp"
 
+#include "Parallel/Printf.hpp"
+
 namespace Parallel {
 /*!
  * \ingroup ParallelGroup
- * \brief Create a converse CmiNodeLock
+ * \brief A typesafe wrapper for the Converse nodelock, with safe creation,
+ * destruction, and serialization.
+ *
+ * \details On construction, this class creates a Converse nodelock, and frees
+ * the lock on destruction. During serialization, the contained nodelock at the
+ * sending node is destroyed to avoid misuse, and throws an error if the lock is
+ * used thereafter. When the lock is deserialized, the nodelock is re-created.
  */
-inline CmiNodeLock create_lock() noexcept { return CmiCreateLock(); }
+class NodeLock {
+ public:
+  NodeLock() noexcept {
+    lock_ = CmiCreateLock();
+  }
 
-/*!
- * \ingroup ParallelGroup
- * \brief Free a converse CmiNodeLock. Using the lock after free is undefined
- * behavior.
- */
-inline void free_lock(const gsl::not_null<CmiNodeLock*> node_lock) noexcept {
+  explicit NodeLock(CkMigrateMessage* /*message*/) noexcept {}
+
+  NodeLock(const NodeLock&) = delete;
+  NodeLock& operator=(const NodeLock&) = delete;
+  NodeLock(NodeLock&& moved_lock) noexcept {
+    moved_lock.destroy();
+    lock_ = CmiCreateLock();
+  }
+
+  NodeLock& operator=(NodeLock&& moved_lock) noexcept {
+    moved_lock.destroy();
+    lock_ = CmiCreateLock();
+    return *this;
+  }
+
+  ~NodeLock() noexcept { destroy(); }
+
+  void lock() noexcept {
+  if (UNLIKELY(destroyed_)) {
+    ERROR("Trying to lock a destroyed lock");
+  }
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wold-style-cast"
-  CmiDestroyLock(*node_lock);
+    CmiLock(lock_);
 #pragma GCC diagnostic pop
 }
 
-/*!
- * \ingroup ParallelGroup
- * \brief Lock a converse CmiNodeLock
- */
-inline void lock(const gsl::not_null<CmiNodeLock*> node_lock) noexcept {
+  bool try_lock() noexcept {
```

Comment:
can you use `try_lock` to do a very rough test of this class? And maybe test that you hit the error in one case. I don't think you need to test that every function errors, basically just that if the thing is serialized the lock is in a `destroyed` state. What do you think?